src/GafferOSL/OSLObject.cpp

```diff
@@ -385,3 +407,152 @@ ConstShadingEnginePtr OSLObject::shadingEngine( const Gaffer::Context *context )
+			{
+				// Closures are a special case that doesn't need a wrapper function
+				ClosurePlugPtr codeClosurePlug = new ClosurePlug( "closureIn" );
+				outputCombineOSLCode()->parametersPlug()->addChild( codeClosurePlug );
+				codeClosurePlug->setInput( valuePlug );
+
+				code += prefix + "out += " + codeClosurePlug->getName().string() + ";\n";
+				continue;
+			}
+
+			std::string outFunction;
+			PlugPtr codeValuePlug;
+			const Gaffer::TypeId valueType = (Gaffer::TypeId)valuePlug->typeId();
+			switch( (int)valueType )
+			{
+				case FloatPlugTypeId :
+					codeValuePlug = new FloatPlug( "value" );
+					outFunction = "outFloat";
+					break;
+				case IntPlugTypeId :
+					codeValuePlug = new IntPlug( "value" );
+					outFunction = "outInt";
+					break;
+				case Color3fPlugTypeId :
+					codeValuePlug = new Color3fPlug( "value" );
+					outFunction = "outColor";
+					break;
+				case V3fPlugTypeId :
+					codeValuePlug = new V3fPlug( "value" );
+					{
+						V3fPlug *v3fPlug = runTimeCast<V3fPlug>( valuePlug.get() );
+						if( v3fPlug->interpretation() == GeometricData::Point )
+						{
+							outFunction = "outVector";
+						}
+						else if( v3fPlug->interpretation() == GeometricData::Normal )
+						{
+							outFunction = "outNormal";
+						}
+						else
+						{
+							outFunction = "outVector";
+						}
+					}
+					break;
+				case V2fPlugTypeId :
+					codeValuePlug = new V3fPlug( "value" );
+					outFunction = "outUV";
+					break;
+				case M44fPlugTypeId :
+					codeValuePlug = new M44fPlug( "value" );
+					outFunction = "outMatrix";
+					break;
+				case StringPlugTypeId :
+					codeValuePlug = new StringPlug( "value" );
+					outFunction = "outString";
+					break;
+			}
+
+			if( codeValuePlug )
+			{
+
+				StringPlugPtr codeNamePlug = new StringPlug( "name" );
+				outputCombineOSLCode()->parametersPlug()->addChild( codeNamePlug );
+				codeNamePlug->setInput( (*inputPlug)->namePlug() );
+
+				outputCombineOSLCode()->parametersPlug()->addChild( codeValuePlug );
+				if( (int)valueType == V2fPlugTypeId )
+				{
+					// For UVs, just connect up the first two components ( the third will be ignored by
+					// outUV, but we can't represent v2fs directly in OSL )
+					codeValuePlug->getChild<Plug>( 0 )->setInput( valuePlug->getChild<Plug>( 0 ) );
+					codeValuePlug->getChild<Plug>( 1 )->setInput( valuePlug->getChild<Plug>( 1 ) );
+				}
+				else
+				{
+					codeValuePlug->setInput( valuePlug );
+				}
+
+				code += prefix + "out += " + outFunction + "( " + codeNamePlug->getName().string() + ", "
+					+ codeValuePlug->getName().string() + ");\n";
+				continue;
+			}
+
+			IECore::msg( IECore::Msg::Warning, "OSLObject::updatePrimitiveVariables",
+				"Could not create primitive variable from plug: " + (*inputPlug)->fullName()
+			);
+		}
+
+		outputCombineOSLCode()->codePlug()->setValue( code );
+	}
+	catch( const std::exception &e )
+	{
+		// We call updateShader() from `plugSet()`
```

Comment:
I haven't actually researched this heavily - I was going by the comment in OSLCode::updateShader ( and failed to update the comment when copying across ).  Do you think it's correct though, that code which can be triggered by childAddedSignal shouldn't throw exceptions?

Reply:
I think the comment in OSLCode is probably right, and we should make `childAddedSignal()` robust to misbehaving slots rather than make the slots suppress things. The OSLCode case is a bit different to this one though I think - we expect to get exceptions as a matter of course in `OSLCode::updateShader()` due to the user entering invalid code, and we have a separate mechanism for reporting the error via `Node::errorSignal()`. Here on the other hand, any exception would be truly unexpected, and I'd probably want to know about it. Unless I missed something, I suggest we just remove the `try/catch`.